File: java/src/P05Functor.java
import java.util.Optional;
import java.util.function.Function;

public class P05Functor {
    public static void main(String[] args) {
        var some2 = Optional.of(2);
        assert Optional.of(4).equals(some2.map(x -> f.apply(x)));

        var none = Optional.<Integer>empty();
        assert Optional.empty().equals(none.map(x -> f.apply(x)));

        assert Optional.of(4).equals(OPTION_FUNCTOR.fmap(f, some2));
        assert Optional.empty().equals(OPTION_FUNCTOR.fmap(f, none));
    }

    private final static Function<Integer, Integer> f = x -> x + 2;
    private final static OptionFunctor OPTION_FUNCTOR = new OptionFunctor();
}

interface Functor {
    <A, B> Optional<B> fmap(Function<A, B> f, Optional<A> fa);
}

class OptionFunctor implements Functor {
    @Override
    public <A, B> Optional<B> fmap(Function<A, B> f, Optional<A> fa) {
        if (!fa.isPresent()) return Optional.empty();
        else return Optional.of(f.apply(fa.get()));
    }
}

File: kotlin/src/P05Functor.kt
import java.util.*
import kotlin.test.assertEquals

object P05Functor {
    @JvmStatic
    fun main(args: Array<String>) {
        val some2 = Optional.of(2)
        assertEquals(Optional.of(4), some2.map { x -> f(x) })

        val none = Optional.empty<Int>()
        assertEquals(Optional.empty(), none.map { x -> f(x) })

        assertEquals(Optional.of(4), OptionFunctor.fmap(f, some2))
        assertEquals(Optional.empty(), OptionFunctor.fmap(f, none))
    }

    val f: (Int) -> Int = { x -> x + 2 }
}

interface Functor {
    fun <A, B> fmap(f: (A) -> B, fa: Optional<A>): Optional<B>
}

object OptionFunctor : Functor {
    override fun <A, B> fmap(f: (A) -> B, fa: Optional<A>): Optional<B> {
        return when {
            !fa.isPresent -> Optional.empty()
            else -> Optional.of(f(fa.get()))
        }
    }
}